Toplean_Alexandru/week-02/Domain/Domain/CreateRestauratOp/CreateRestaurantOp.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Domain.Models;
+using Infrastructure.Free;
+using LanguageExt;
+using LanguageExt.ClassInstances.Pred;
+using static Domain.Domain.CreateRestauratOp.CreateRestaurantResult;
+using static Domain.Models.Restaurant;
+
+namespace Domain.Domain.CreateRestauratOp
+{
+    public class CreateRestaurantOp : OpInterpreter<CreateRestaurantCmd, CreateRestaurantResult.ICreateRestaurantResult, Unit>
+    {
+        public override Task<ICreateRestaurantResult> Work(CreateRestaurantCmd Op, Unit state)
+        {
+            //validate
+
+            return Exists(Op.Name) ?
```

Comment:
Would it be fair to assume that most of these validations belong to the actual command? 

From my point of view the command itself should contain the validation logic which should be called from the operation itself. 

Please be aware that that does not include the !"Exists" preconditions. That precondition checks that a given restaurant exists in your environment, which is not in the ownership of the command. 

Can you try a refactoring to clearly separate the responsibilities?